### ISSUE TITLE
Adding a `lazy val` to a `trait` emits backwards-compatible code with wrong semantics

### ISSUE THREAD

**Issue by armanbilge** [lampepfl/dotty#16352]
## Compiler version

`3.3.0-RC1-bin-20221115-e587a81-NIGHTLY`

## Minimized code

**tl;dr** `sbt app/runMain Bar` in https://github.com/armanbilge/sandbox/commit/345b785eac4d6a873b7e05e6a711da048dbf917b

1. Compile old `Foo`.

```scala
trait Foo {
  def main(args: Array[String]): Unit = ()
}
```

2. Compile `Bar`.
```scala
object Bar extends Foo
```

3. Compile new `Foo`.
```scala
trait Foo {
  lazy val sameSame: AnyRef = new AnyRef

  def main(args: Array[String]): Unit = println(sameSame eq sameSame)
}
```

## Output

4. Run `Bar` with new `Foo`.
```
sbt:sandbox> app/runMain Bar
[info] running Bar 
false
```

## Expectation

Lazy val semantics dictate that `sameSame eq sameSame` should always be `true`.

## Analysis

If we look at the decompiled bytecode for new `Foo`, we see it has emitted a `default public Object sameSame()` method which it is using as an accessor. However, because `Bar` was compiled against an old `Foo`, it is not overriding that method. Therefore the `lazy val` effectively has the semantics of a `def`.

It seems like that should not be a `default` method, since the default implementation does not have the correct semantics.

```scala
import scala.Predef$;
import scala.runtime.BoxesRunTime;

public interface Foo {
    public static void $init$(Foo $this) {
    }

    public static Object sameSame$(Foo $this) {
        return $this.sameSame();
    }

    default public Object sameSame() {
        return new Object();
    }

    public static void main$(Foo $this, String[] args) {
        $this.main(args);
    }

    default public void main(String[] args) {
        Predef$.MODULE$.println((Object)BoxesRunTime.boxToBoolean((this.sameSame() == this.sameSame() ? 1 : 0) != 0));
    }
}
```

## Credits

@s5bug for stumbling on this by way of https://github.com/bkirwi/decline/issues/461.

## See also

- https://github.com/scala/bug/issues/12692

**Comment by odersky** [comment]
I _think_ that's as expected. Changing a trait requires recompilation of extending classes in some cases.

**Comment by odersky** [comment]
Yes, it looks to me that MiMa should flag it as a backwards incompatible change. /cc @dwijnand to confirm.
